### PR TITLE
working text detection (specifically set up for locations)

### LINKED ISSUE #3: recognize location names
(self-explanatory)

ideas:
obviously char data is used as the needle image in templatematching
crop haystack image to relevant area before feeding into templatematching
take output arrays for all characters and build an array of indices corresponding to each character
every pixel will have the closest matching character's corresponding index, then assign the best match and mark occupied pixels, then do nearby match in unoccupied pixels (or just do the next best match in unoccupied pixels--try both approaches and see what performs better)

### PR DESCRIPTION
**Pull request by jahndan**
(resolves #3)
in theory this should work for all text detection we'll need to do, but there may be a few mistakes we'll find with multi-line text, etc.

obviously, right now this only reads location names and prints them to the console (when we have actual tracker logic set up, they should ideally trigger side-effects and update the tracker state)

this was rebased multiple times on top of branch `font-updates` so there may be notes/other artifacts (i.e. comments that clutter some files) leftover that are no longer needed (please point this out in review if so)

### CODE REVIEW

File: src/en_model.py [hunk 1]
```diff
@@ -1,4 +1,144 @@
...
+lower_alpha = set("abcdefghijklmnopqrstuvwxyz")
+upper_alpha = set("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+alphabet = lower_alpha | upper_alpha
+numbers = set("0123456789")
+# do we ever need to identify punctuation?
+# punctuation = set("$!?,.•/‘’“”„«»()♂♀+-*#=&~:;")
+# ellipsis … may not be used in english version of game
+
+# char dicts for each part of the logic (é/e doesn't distinguish anything)
+# locations = sorted(
```
Comment (AitanSingh) [inline src/en_model.py]: Is this necessary here?

**Review by AitanSingh — COMMENTED** [review]
Looked good, not that much clutter although would be worth going through the TODOs and deleting any that have already been completed

File: src/en_model.py (hunk 1 above)
Comment (jahndan) [reply src/en_model.py]: oops that's a rebase fragment i forgot to remove

**Comment by jahndan** [comment]
test case for multi-row text works (archived here in case it's useful)

```py
from en_model import locations, locations_palette
from en_fontmap import normal_fontmap, bold_fontmap
from font import palette_transfer

helloworld = palette_transfer(
    set("HelloWorld"), normal_fontmap, locations_palette
)
W, H, d, e, o, r, l = [img for sort, char, img, mask in helloworld]
HW = numpy.pad(
    numpy.concatenate(
        [
            numpy.concatenate(
                [H, e, l, l, o, numpy.full((16, 2, 3), 255, dtype=numpy.uint8)],
                axis=1,
            ),
            numpy.concatenate([W, o, r, l, d], axis=1),
        ],
        axis=0,
    ),
    ((0, 0), (4, 4), (0, 0)),
    constant_values=255,
)
print("original shape", HW.shape)
opencv.imshow("original", HW)
print(model.process_text(HW, locations))
```

**Pull request merged**